Lesson 5

ilya-burinskiy opened the pull request.

Commit:
Answers correct scope

Commit:
Category defautl scope ordering by name

Commit:
Replaced class method list_titles_by_category by scope

Commit:
Using scope method in started_tests_by_level

Commit:
Added level-title index

Commit:
Firt name, last name, email validation

Commit:
Answer body validation

Commit:
Category name validation

Commit:
Test title and level validations

Commit:
Question answer count validation

Commit:
Answer count validation fix

VidgarVii commented on app/models/test.rb:
```diff
@@ -3,11 +3,22 @@ class Test < ApplicationRecord
+  scope :medium, -> { by_level(2..4) }
+  scope :difficult, -> { by_level(5..Float::INFINITY) }
+
+  scope :list_titles_by_category, lambda { |pattern|
```
scope всегда должен возвращать ActiveRecord Relation. Этот метод не надо трогать

VidgarVii commented on app/models/question.rb:
```diff
@@ -1,4 +1,15 @@
   has_many :answers, dependent: :destroy
+
+  validates :title, presence: true
+  validate :validate_answer_count, on: :create
```
нам надо создание ответа для вопроса валидировать, а не сам вопрос)

VidgarVii commented on app/models/question.rb:
```diff
@@ -1,4 +1,15 @@
+  private
+
+  def validate_answer_count
+    if answers.count < 1 || answers.count > 4
```
можно 1 условием обойтись

Commit:
list_titles_by_category is Test class method

Commit:
Adding answer to question validation

Commit:
Using one condition in answer count validation

Commit:
Question validation fix

Question is validated when the number of
answers is zero

Commit:
ru.yml

Commit:
seed updated

VidgarVii commented on db/seeds.rb:
```diff
@@ -27,82 +27,83 @@
 users_seeds.each do |seed|
   User.find_or_create_by(first_name: seed.first_name,
```
find_or_create_by!

VidgarVii commented on app/models/question.rb:
```diff
@@ -1,4 +1,20 @@
+
+  private
+
+  def validate_answer_count
```
эта валидация тут вообще не нужна. Или у тебя есть какой-то тайный замысел?

VidgarVii commented on db/seeds.rb:
```diff
@@ -27,82 +27,83 @@
-author_id = User.first.id
+author = User.first
 tests_seeds = [
   Test.new(title: 'Multiple inheritance', level: 2,
```
Если использовать классический вариант, как предлагает дока, будет на много читабильней.
Test.create!([{...}, {…}])

Commit:
Renamed question title to question body

Commit:
Using create! instead of find_or_create_by!

Commit:
Question validation fix

- Deleted validate_answer_count
- Deleted should_validate?
- Added body uniqueness

Commit:
Added answer body uniqueness validation

Commit:
Added User email uniqueness validation

Commit:
Category name uniqueness validation

Commit:
Test cannot have more than one user whith same id

Commit:
Indexes added to Answer and Question

Commit:
Answer validation fix

- Added on: :create to validate_add_to_question
- Added body uniqueness: { scope: :question }

Commit:
Question validation fix

- Added body uniqueness: { scope: :test }

Commit:
TestsStartedByUser validation fix

Removed user_id uniqueness: { scope: :test_id }

Commit:
Seed data only if tables does not exist

Commit:
Seed data only if tables does not exist

Commit:
Removed multiple validation per one attribute

Commit:
Only one use validates for :level

Commit:
Merge remote-tracking branch 'origin/lesson-5' into lesson-5
Случайно запушил раньше времени

VidgarVii commented on app/models/answer.rb:
```diff
@@ -1,3 +1,14 @@
 class Answer < ApplicationRecord
   belongs_to :question
+  validates :body, uniqueness: { scope: :question }, presence: true
+  validate :validate_add_to_question, on: :create
```
чтобы было понятней. я бы обозвал валидатор как то "максимальное кол-во ответов для вопроса"

VidgarVii commented on app/models/answer.rb:
```diff
@@ -1,3 +1,14 @@
 class Answer < ApplicationRecord
   belongs_to :question
+  validates :body, uniqueness: { scope: :question }, presence: true
```
belongs_to и validates  логически разные вещи, которые надо разделять пустыми строками
https://github.com/rubocop-hq/rails-style-guide#macro-style-methods

VidgarVii commented on app/models/user.rb:
```diff
@@ -3,12 +3,13 @@ class User < ApplicationRecord
+  validates :first_name, presence: true
+  validates :last_name, presence: true
+  validates :email, uniqueness: true, presence: true,
+                    format: { with: /[A-Za-z0-9+_.-]+@[A-Za-z0-9.-]+/,
```
можно воспользоваться готовым паттерном URI::MailTo::EMAIL_REGEXP

Commit:
Small fixes
